### PR TITLE
[WFCORE-5740] Add the ability to ignore unavailable realms for a dist…

### PR DESCRIPTION
**Pull request by lvydra**
https://issues.redhat.com/browse/EAP7-1862
https://issues.redhat.com/browse/WFCORE-5740
https://issues.redhat.com/browse/ELY-2305

former closed PR: https://github.com/wildfly/wildfly-core/pull/4974

### CODE REVIEW

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties
Comment (OndrejKotek) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties:992]: IIUIC, this is a bit imprecise. It should be clear that the event is emitted just in case `ignore-unavailable-realms` is enabled.

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties
Comment (OndrejKotek) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties:991]: This is imprecise. It should be clear that the authn process fails by default when an unavailable realm is reached and this attribute is false.

File: elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java
Comment (OndrejKotek) [inline elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java:738]: > DistributedRealmOne

Can you use more descriptive names please? For example DistributedRealmNoUnavailable, DistributedRealmFirstUnavailableFailing, DistributedRealmFirstUnavailableIgnored.

File: elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java
Comment (OndrejKotek) [inline elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java:755]: Effect of `emit-events` is not tested.

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties
Comment (OndrejKotek) [reply elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: I suggest to also clarify that an exception is thrown.

The descriptions in the XSD should be also updated.

**Comment by lvydra** [comment]
Hi @fjuma @Skyllarr, could I ask you for a review of this PR?

File: elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java
Comment (Skyllarr) [inline elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java:160]: @lvydra just a minor but we should add a transformer test as well to SubsystemTransformersTestCase. An example of such test can bee seen in this commit for example https://github.com/wildfly/wildfly-core/commit/13e3181d4528a6b74c09792725edfac382d09c00

File: elytron/src/main/resources/schema/wildfly-elytron_16_0.xsd
Comment (fjuma) [reply elytron/src/main/resources/schema/wildfly-elytron_16_0.xsd:1166]: Thanks, @Skyllarr! @lvydra I will submit a PR for the schema bump to 17 and will add the link to it here either today or Monday.

File: elytron/src/main/java/org/wildfly/extension/elytron/DistributedRealmDefinition.java
Comment (Skyllarr) [inline elytron/src/main/java/org/wildfly/extension/elytron/DistributedRealmDefinition.java:132]: @lvydra jsut a minor, should here be 

```
emitEvents && ignoreUnavailableRealms ? (realmIndex) -> ...
```

so that the emitting callback is registered only when the unavailable realm should be ignored?

File: elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java
Comment (Skyllarr) [inline elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java:761]: @lvydra  this assertion fails for me locally when running `mvn clean install -Dtest=RealmsTestCase` . It is because the target/audit.log is empty with no lines during execution of the test. But after the test finishes the file gets written and I can see the event being logged there.
Does it work locally for you?

File: elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java
Comment (Skyllarr) [inline elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java:768]: @lvydra Just a minor but can we check the name of the realm that threw the SecurityRealmUnavailableEvent as well just to know that the message is related to the event we want to test?

File: elytron/src/main/resources/schema/wildfly-elytron_16_0.xsd
Comment (fjuma) [reply elytron/src/main/resources/schema/wildfly-elytron_16_0.xsd:1166]: @lvydra Here's the PR for subsystem bump:

https://github.com/wildfly/wildfly-core/pull/5463

Please update this PR so it's based off of that one. Thanks!

File: elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java
Comment (fjuma) [reply elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java:774]: @lvydra Just to check, has this comment from @OndrejKotek been addressed yet?

Comment (lvydra) [reply elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java:774]: @fjuma Yes, this one has already been addressed.

File: elytron/src/main/resources/schema/wildfly-elytron_16_0.xsd
Comment (lvydra) [reply elytron/src/main/resources/schema/wildfly-elytron_16_0.xsd]: @Skyllarr @fjuma updated

File: elytron/src/main/java/org/wildfly/extension/elytron/DistributedRealmDefinition.java
Comment (lvydra) [reply elytron/src/main/java/org/wildfly/extension/elytron/DistributedRealmDefinition.java]: @Skyllarr logic on Elytron side is taking care of that, however, it seems better to add it here too.

File: elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java
Comment (lvydra) [reply elytron/src/test/java/org/wildfly/extension/elytron/RealmsTestCase.java]: @Skyllarr should be fixed now.

**Comment by lvydra** [comment]
Hi @Skyllarr, thanks for the review, your comments should be addressed now.

**Review by Skyllarr — APPROVED** [review]

**Comment by Skyllarr** [comment]
@lvydra Thank you!

**Review by fjuma — APPROVED** [review]
Thanks @lvydra!

**Review by OndrejKotek — APPROVED** [review]

**Comment by yersan** [comment]
@lvydra notice this is going to conflict with #5463 since the commit hashes of similar changes are different, to avoid it, you could rebase on top of it, or just wait until it gets merged and drop the commit from your PR.

**Comment by darranl** [comment]
@lvydra +1 to the comment from @yersan by changing the commit for the version bump this is introducing merge conflicts when combined with the other PRs as git sees it as a different commit.

**Review by darranl — DISMISSED** [review]
Please don't change the commits from the version bump PR, the exact commits with exact same SHA need to be used.

**Comment by lvydra** [comment]
Thanks @yersan @darranl, should be rebased now.

**Comment by fjuma** [comment]
Thanks @lvydra! 

@yersan @darranl Just FYI, the commit SHAs are correct now.

**Comment by darranl** [comment]
Please don't make any further changes to this PR - I am going to combine it with a couple of other PRs so we can get them merged together.

Thank you for this implementation.

**Pull request merged**
